OpTestHMC/OpTestGSBStaticKey : align secureboot check-APIs

rohan-r-deshpande opened the pull request:
OpTestHMC is added with all the check_APIs related to secureboot testing. This commit addresses the following API additions -
1. check_hmc_level_secureboot_state()
2. check_grub_signature()
3. get_grub_file()

OpTestGSBStaticKey is changed in a way that the above mentioned APIs have been moved to the common area so as to provide generic availibility towards wider usage.

rohan-r-deshpande commented:
Detailed log : [sb_check_APIs_log.txt](https://github.com/open-power/op-test/files/11498794/sb_check_APIs_log.txt)

harihare commented on common/OpTestHMC.py:
```diff
@@ -80,6 +97,41 @@ def check_kernel_signature(self):
+            if not grub_filename:
+                self.fail("%s: Failed to get grub file" % (self.distro))
+        else:
+            raise self.skipTest("Test currently supported on SLES and RHEL releases")
```
I think statment is wrong here, when it is supported why do we skip it.

rohan-r-deshpande replied:
Right, ideally I meant it to be - _"Test currently supported on SLES & RHEL realeases **only**"_
Will change it.

harihare commented on common/OpTestHMC.py:
```diff
@@ -59,11 +59,28 @@ class OpSecureBootUtilities():
+        output = self.hmc_con.run_command(cmd, timeout=300)
+        if int(output[0]) == 2: # Value '2' means Secure Boot enabled
+            return True
+        elif int(output[0]) == 0: # Value '0' means Secure Boot disabled
```
is this check required, is there any other states other then enalbed or disabled.

rohan-r-deshpande replied:
Yes, as of now only two states exist. However, there could be a possibility of one more state in the future. Hence the conditions.

harihare commented on testcases/OpTestGSBStaticKey.py:
```diff
@@ -128,86 +120,59 @@ def getPRePDisk(self):
-            out = self.c.run_command("pbl --install || echo 'no'")
-            if "no" in "".join(out):
-                self.fail("SLES: Failed to disable Secure Boot")
+        if self.kernel_signature == "signed_kernel":
```
What is the difference between old and new implementation. Does the new implemention work for NV environment too.

rohan-r-deshpande replied:
> _What is the difference between old and new implementation._ 
The flow remains the same, but the API for checking the kernel-signature has been moved to a different location. Hence, instead of failing the test from the API, now the flow is handled in a way which would return the signature and based on that we would FAIL the test here in this file.

> _Does the new implemention work for NV environment too._
Not sure what is meant by NV environment?

sacsant commented on testcases/OpTestGSBStaticKey.py:
```diff
@@ -128,86 +120,59 @@ def getPRePDisk(self):
+                    # Running grub2-install on PReP disk
+                    out = self.c.run_command("grub2-install %s" % self.prepDisk)
+                    if "Installation finished. No error reported." not in out:
+                        self.fail("RHEL: Failed to install on PReP partition")
```
In case of a failure, should we restore the grub2 to original state. If we simply fail the testcase, will reboot of system leave it in un-bootable state?

rohan-r-deshpande replied:
Not sure on this part, needs to be tested.

rohan-r-deshpande replied:
Done.

sacsant commented on testcases/OpTestGSBStaticKey.py:
```diff
@@ -128,86 +120,59 @@ def getPRePDisk(self):
+                    out = self.c.run_command("dd if=%s of=%s || echo 'no'" % 
+                                             (self.grub_filename, self.prepDisk))
+                    if "no" in "".join(out):
+                        self.fail("RHEL: Failed to copy to PReP partition")
```
Same here. In case of a failure, should we restore the grub2 to original state. If we simply fail the testcase, will reboot of system leave it in un-bootable state?

PraveenPenguin replied:
yes good idea Sachin ,

rohan-r-deshpande replied:
Not tested in this way.

Well, to test it - how can this scenario be reproduced?
1] Introducing a failure in the PreP copying.
2] Then in case of a failure - restoring the grub2 to the original state - How is it done?

rohan-r-deshpande replied:
Done

sacsant commented on testcases/OpTestGSBStaticKey.py:
```diff
@@ -128,86 +120,59 @@ def getPRePDisk(self):
+                self.c.run_command(cmd)
+                out = self.c.run_command("pbl --install || echo 'no' ")
+                if "no" in "".join(out):
+                    self.fail("SLES: Failed to enable Secure Boot")
```
Similar comments for SLES as well

rohan-r-deshpande replied:
@sacsant 
- As discussed with Sastry, from SLES15-sp5 onwards, Secureboot is by-default in an enabled state at the OS level and can be checked through the "yast2 bootloader". 
- This code was written earlier, and hence here onwards is not required to be handled specifically for SLES. Hence going ahead and removing the same.

PraveenPenguin commented:
@rohan-r-deshpande any update here ?

rohan-r-deshpande commented:
> @rohan-r-deshpande any update here ?

Have addressed the review comments and here are the updated logs - 
RHEL - [sb_checkAPI_rhel_runlog.txt](https://github.com/open-power/op-test/files/11749645/sb_checkAPI_rhel_runlog.txt)

SUSE - [sb_checkAPI_sles_runlog.txt](https://github.com/open-power/op-test/files/11749648/sb_checkAPI_sles_runlog.txt)

nasastry commented on common/OpTestHMC.py:
```diff
@@ -80,6 +97,41 @@ def check_kernel_signature(self):
+        '''
+        Fetch the grub-file based on the distro
+        '''
+        if self.distro == 'rhel':
```
This function code can be reduced:
if self.distro == 'rhel':
    rpm_name = 'grub2-ppc64le'
    filename = 'core.elf'
if self.distro == 'suse':
    rpm_name = 'grub2-powerpc-ieee1275'
    filename = 'grub.elf'

now run the command and check for the file existence.

nasastry commented on common/OpTestHMC.py:
```diff
@@ -80,6 +97,41 @@ def check_kernel_signature(self):
+        cmd = "strings %s | tail -1" % grub_filename
+        output = self.console.run_command(cmd)
+        if "Module signature appended" in output[0]:
+            return "signed_grub"
```
Why not return 'true' or 'false'?

sacsant replied:
Yup and also how about changing the name to is_grub_signed() and return true/false?

sacsant commented on common/OpTestHMC.py:
```diff
@@ -59,19 +59,36 @@ class OpSecureBootUtilities():
         self.console = console
         self.distro = distro
 
+    def check_hmc_level_secureboot_state(self):
```
I guess reference to HMC is to convey that we are using HMC to check the value of secure boot. In that case how about check_lpar_secureboot_state?

rohan-r-deshpande replied:
Done

sacsant commented on common/OpTestHMC.py:
```diff
@@ -59,19 +59,36 @@ class OpSecureBootUtilities():
 
+    def check_hmc_level_secureboot_state(self):
+        '''
+        Check whether the secure-boot is enabled and HMC level.
```
and HMC level can be remove from this line. It does not convey anything and indicates an incomplete sentence.

rohan-r-deshpande replied:
Done

abdhaleegit commented:
@rohan-r-deshpande As discussing in slack

if I see the class methods, they can still fit in existing HMCUtils class

the reason is OpTestHMC.py  will have all methods performing HMC related operations
get_grub_file
check_grub_sign
check_kernel_signature
all 3 are linux os operations
check_hmc_level_secureboot_state   is HMC operation .. so this fits in class HMCUtil()

another quesiton, do you want mkvterm console to run these commands ? or it can be simple ssh commands
get_grub_file
check_grub_sign
check_kernel_signature

So  by not having a new class   you can always  from test reuse the class methods
self.cv_HMC.check_hmc_level_secureboot_state()
self.util.get_grub_file()
self.util.check_grub_sign()
self.utils.check_kernel_signature()

but if you feel still wants all secureboot related utilites at one place.. you can have a separate util out of hmc

so if you see OpTestUtils.py  it has all Linux related commands/operations

abdhaleegit commented on testcases/OpTestGSBStaticKey.py:
```diff
@@ -249,7 +210,9 @@ def collectData(self):
         # Check the hmc secure boot state
-        hmc_secureboot = self.check_hmc_secureboot_state()
+        sb_utilities = OpSecureBootUtilities(self.cv_HMC, self.hmc_con)
+        hmc_secureboot = sb_utilities.check_hmc_level_secureboot_state()
```
if moved to existing class of HMCUtils than this is just a simple call

self.cv_HMC.check_hmc_level_secureboot_state()

abdhaleegit reviewed (changes requested):
Related to above discussion

abdhaleegit commented:
any update on the latest code ?

rohan-r-deshpande commented:
Update execution log -

[sb_checkAPI_rhel89_runlog.txt](https://github.com/open-power/op-test/files/12321692/sb_checkAPI_rhel89_runlog.txt)
[sb_checkAPI_rhel93_runlog.txt](https://github.com/open-power/op-test/files/12321699/sb_checkAPI_rhel93_runlog.txt)

abdhaleegit reviewed (approved):
LGTM

abdhaleegit commented:
@rohan-r-deshpande Thanks for working on all the review comments
